cmd/thv-operator/controllers/virtualmcpserver_controller.go

```diff
@@ -0,0 +1,1071 @@
+// and uses the same annotation constant for consistency.
+func (r *VirtualMCPServerReconciler) getVmcpConfigChecksum(
+	ctx context.Context,
+	vmcp *mcpv1alpha1.VirtualMCPServer,
+) (string, error) {
+	if vmcp == nil {
+		return "", fmt.Errorf("vmcp cannot be nil")
+	}
+
+	configMapName := fmt.Sprintf("%s-vmcp-config", vmcp.Name)
+	configMap := &corev1.ConfigMap{}
+	err := r.Get(ctx, types.NamespacedName{
+		Name:      configMapName,
+		Namespace: vmcp.Namespace,
+	}, configMap)
+
+	if err != nil {
+		// Preserve error type for IsNotFound checks
+		return "", fmt.Errorf("failed to get vmcp Config ConfigMap %s/%s: %w",
+			vmcp.Namespace, configMapName, err)
+	}
+
+	// Use the standard checksum annotation constant for consistency
+	checksumValue, ok := configMap.Annotations[checksum.ContentChecksumAnnotation]
+	if !ok {
+		return "", fmt.Errorf("vmcp Config ConfigMap %s/%s missing %s annotation",
+			vmcp.Namespace, configMapName, checksum.ContentChecksumAnnotation)
+	}
+
+	if checksumValue == "" {
+		return "", fmt.Errorf("vmcp Config ConfigMap %s/%s has empty %s annotation",
+			vmcp.Namespace, configMapName, checksum.ContentChecksumAnnotation)
+	}
+
+	return checksumValue, nil
+}
+
+// ensureDeployment ensures the Deployment exists and is up to date
+func (r *VirtualMCPServerReconciler) ensureDeployment(
+	ctx context.Context,
+	vmcp *mcpv1alpha1.VirtualMCPServer,
+) (ctrl.Result, error) {
+	ctxLogger := log.FromContext(ctx)
+
+	// Fetch vmcp Config ConfigMap checksum to include in pod template annotations
+	vmcpConfigChecksum, err := r.getVmcpConfigChecksum(ctx, vmcp)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			ctxLogger.Info("vmcp Config ConfigMap not found yet, will retry",
+				"vmcp", vmcp.Name, "namespace", vmcp.Namespace)
+			return ctrl.Result{RequeueAfter: 5 * time.Second}, nil
+		}
+		ctxLogger.Error(err, "Failed to get vmcp Config checksum")
+		return ctrl.Result{}, err
+	}
+
+	deployment := &appsv1.Deployment{}
+	err = r.Get(ctx, types.NamespacedName{Name: vmcp.Name, Namespace: vmcp.Namespace}, deployment)
+
+	if errors.IsNotFound(err) {
+		dep := r.deploymentForVirtualMCPServer(ctx, vmcp, vmcpConfigChecksum)
+		if dep == nil {
+			return ctrl.Result{}, fmt.Errorf("failed to create Deployment object")
+		}
+		ctxLogger.Info("Creating a new Deployment", "Deployment.Namespace", dep.Namespace, "Deployment.Name", dep.Name)
+		if err := r.Create(ctx, dep); err != nil {
+			ctxLogger.Error(err, "Failed to create new Deployment")
+			// Record event for deployment creation failure
+			if r.Recorder != nil {
+				r.Recorder.Eventf(vmcp, corev1.EventTypeWarning, "DeploymentCreationFailed",
+					"Failed to create Deployment: %v", err)
+			}
+			return ctrl.Result{}, err
+		}
+		// Record event for successful deployment creation
+		if r.Recorder != nil {
+			r.Recorder.Event(vmcp, corev1.EventTypeNormal, "DeploymentCreated",
+				"Deployment created successfully")
+		}
+		return ctrl.Result{RequeueAfter: 0}, nil
```

Comment:
question: why do we need to requeue if we've already created the deployment and things are up to date?